Add Key Verification to uProxy

mollyling opened the pull request:
In parallel to this change is the set of necessary changes to uproxy-lib, which are in-review here: https://github.com/uProxy/uproxy-lib/pull/418.  But as uproxy-lib's getting merged into here at the same time, when uproxy-lib#418 is reviewed, I'll put the changes into uproxy's copy of it.  Then I can rebase against master.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/2459)
<!-- Reviewable:end -->

bemasc commented:
Review status: 0 of 18 files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[src/generic_core/key-verify.ts, line 1 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlsYzn4Pqz5uOV6oDX:-KJlsYzoIuCt8-DGFFjY:787296444) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L1)):*
> ```XML
> import crypto = require('crypto');
> ```

Do we know how much stuff this pulls into the uProxy build?  If it's just the ciphers we need that's fine, but if it's also RSA and lots of other stuff ... might be better to use a purpose-built hash package.  Also, we're already importing an md5 package for the identicons, so if this introduces another md5 implementation then we should consider getting rid of the other one.

---

*[src/generic_core/key-verify.ts, line 12 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlrOfgWW6OF05Weduo:-KJlrOfhDG80EyFArrmC:2128041853) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L12)):*
> ```XML
> 
>   // Send this message to the peer KeyVerify
>   sendMessage(msg:any) :Promise<void>;
> ```

Can you make this `Object` instead of `any`?

---

*[src/generic_core/key-verify.ts, line 16 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlrZ-_VuRd8yQ7I42i:-KJlrZ-_VuRd8yQ7I42j:-1048793785) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L16)):*
> ```XML
>   // Show this Short Authentication String to the user, and ask them
>   // if it's the same as what their peer sees.  resolve(true) if it
>   // is, resole(false) if not.
> ```

resolve

---

*[src/generic_core/key-verify.ts, line 27 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlrtYoa7vmKGRcIMtL:-KJlrtYoa7vmKGRcIMtM:326561980) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L27)):*
> ```XML
>     constructor(public type: string, public version: string, public h3: string,
>                 public hk: string, public clientVersion: string,
>                 public mac: string){}
> ```

Space between `)` and `{`

---

*[src/generic_core/key-verify.ts, line 31 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlsAeKBQQ0f159WpNL:-KJlsAeKBQQ0f159WpNM:-1367753664) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L31)):*
> ```XML
> 
>   export class CommitMessage {
>     constructor(public type: string, public h2: string, public hk: string,
> ```

Could `type` be a `Type` here instead of a string?

---

*[src/generic_core/key-verify.ts, line 52 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlsE4oVLoKd5EjaNPz:-KJlsE4oVLoKd5EjaNQ-:-1278997085) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L52)):*
> ```XML
>     constructor(
>       public type: Type,
>       public value: HelloMessage|CommitMessage|DHPartMessage|
> ```

Could you make these types subtypes of a `Message` supertype?

---

*[src/generic_core/key-verify.ts, line 102 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJluQ49hGoUSLKfIomY:-KJluQ49hGoUSLKfIomZ:1683419231) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L102)):*
> ```XML
> }
> 
> function hexify(buf:Buffer) {
> ```

Did you see `buf.toString('hex')`? https://nodejs.org/api/buffer.html#buffer_buffers_and_character_encodings

---

*[src/generic_core/key-verify.ts, line 107 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlu8M36RAsG1jxgGJu:-KJlu8M36RAsG1jxgGJv:2100317174) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L107)):*
> ```XML
>     var n :number = buf.readInt8(i);
>     if (n < 0) {
>       n = (-n ^ 0xff) + 1;
> ```

Why not `buf.readUInt8`?

---

*[src/generic_core/key-verify.ts, line 111 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJltywtqnrhU-4Ky9J3:-KJltywujahKK615TAQk:-976520566) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L111)):*
> ```XML
>     var c = n.toString(16);
>     if (c.length < 2) {
>       c = "0" + c;
> ```

Prefer singlequotes

---

*[src/generic_core/key-verify.ts, line 122 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpfoK3nw4ZVov2Gnfh:-KJpfoK3nw4ZVov2Gnfi:-1433142964) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L122)):*
> ```XML
>   var linebuf = name + ":";
>   var remain_size = kBufWidth - linebuf.length;
>   for (var i = 0; i < buf.length; i++) {
> ```

I don't understand everything that's going on here, but it seems like overkill and duplication of code.

---

*[src/generic_core/key-verify.ts, line 148 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpg0V7eZKtTEp530JG:-KJpg0V7eZKtTEp530JH:-1727977628) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L148)):*
> ```XML
> // concatenated before we hash them.
> function unbase64Concat(...args:string[]) :Buffer {
>   let buffers = args.map( (val:string) => {
> ```

No space between `( (`

---

*[src/generic_core/key-verify.ts, line 158 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpqFryig6BPenrPTFD:-KJpqFryig6BPenrPTFE:-1418741737) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L158)):*
> ```XML
> // (Delegate above).  For the side starting the session, to construct,
> // just pass in the peer's public key to verify, and the delegate.
> // For the other side, first parse the message with readFirstMessage()
> ```

This usage model seems pretty weird.  Maybe it would make more sense to have a pair of factory methods for the two roles?

---

*[src/generic_core/key-verify.ts, line 166 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpgN68klkP6PZ7vDCC:-KJpgN68klkP6PZ7vDCD:-317086533) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L166)):*
> ```XML
>   // Only written and read by sendNextMessage(), between the time we
>   // call generate_ and get its promise-return.
>   private queued_generations_: {[type:string]:boolean}; // indexed by Type.
> ```

`queuedGenerations_` in our code style

---

*[src/generic_core/key-verify.ts, line 171 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpgTfJwGIFvW_lc_XK:-KJpgTfJwGIFvW_lc_XL:1668718034) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L171)):*
> ```XML
>   // shouldn't determine who's sending 'Commit', but right now, we only support
>   // role_ == 0 sending Commit.
>   private role_:number;
> ```

I'd prefer an enum (or a boolean)

---

*[src/generic_core/key-verify.ts, line 192 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJph1wa4JrBuNFrgveJ:-KJph1wb49texxzS6vYu:1893911013) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L192)):*
> ```XML
>   private totalHash_ :Buffer = null;
> 
>   private static kPgpPassword :string = '';
> ```

I think our code style is `PGP_PASSWORD` for constants.

---

*[src/generic_core/key-verify.ts, line 198 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJphGwy3cc0aX76aczq:-KJphGwy3cc0aX76aczr:855991810) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L198)):*
> ```XML
>   private static seqno :number = 1;
>   private static emptyPreReq :Type[] = [];
>   private static keyMap_ :{[type:string]:string}= {
> ```

`keyMap_` and `prereqMap_` need descriptions.  I presume you're encoding some sort of state machine?  A comma-separated list of ... something ... encoded as a string, is pretty strange.

---

*[src/generic_core/key-verify.ts, line 199 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpiLb84Or9PzV-Glqt:-KJpiLb84Or9PzV-Glqu:-1199597548) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L199)):*
> ```XML
>   private static emptyPreReq :Type[] = [];
>   private static keyMap_ :{[type:string]:string}= {
>     'Hello1':'clientVersion,h3,hk,mac,type,version',
> ```

Instead of repeating string literals, I'd prefer having these keys in some sort of variable or collection, just so we aren't retyping each string in each place it's used and risking a typo.

---

*[src/generic_core/key-verify.ts, line 238 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJphWgLzvQ1l5P08_A7:-KJphWgLzvQ1l5P08_A8:248059513) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L238)):*
> ```XML
>   // conversation.  Useful both for testing and for when this Verifier
>   // is being created in response to a received message.
>   constructor(peerPubKey: string,
> ```

Can collapse this to `private peerPubKey_:string`.  Same with `delegate`.

---

*[src/generic_core/key-verify.ts, line 297 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpjyERT-jbOFBAtR2A:-KJpjyERT-jbOFBAtR2B:-965775546) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L297)):*
> ```XML
>   }
> 
>   public readMessage(msg:any) {
> ```

I find this really hard to read.  I think it would be clearer if each message processing algorithm were split out into its own method, so this method just contains control flow.

I understand why you're trying to keep both sides' flow in a single linear sequence here, but it's fairly counterintuitive to me.  Hopefully it'll be more comprehensible with the detailed logic moved out into helper methods.

---

*[src/generic_core/key-verify.ts, line 307 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpjIVhwn76qR6_5fOt:-KJpjIVhwn76qR6_5fOu:-1146451456) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L307)):*
> ```XML
>       //  2. Remember to keep track of whether the calculation is
>       //     using the base64 or binary values
>       if (type == 'Hello1' || type == 'Hello2') {
> ```

We prefer `===`

---

*[src/generic_core/key-verify.ts, line 412 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpjbEZwqLOmCUPEzBp:-KJpjbE_W0P4aF_VIJYy:-454516842) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L412)):*
> ```XML
>         });
>       } else if (type == 'DHPart2') {
>         // Verify that this is the sam ehk.
> ```

typo?

---

*[src/generic_core/key-verify.ts, line 745 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpmrguNCdN6a8tFWF4:-KJpmrgvtF1mtx36E2Lu:-939994184) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L745)):*
> ```XML
>     if (!this.messages_[Type.DHPart2]) {
>       return this.makeDHPart_(Type.DHPart2).then( (msg:Messages.Tagged) => {
>         // this.set_(msg);  // don't set() unless we've sent or recevied it.
> ```

Commented out code?   Also typo.

---

*[src/generic_core/key-verify.ts, line 786 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpnOGYNfyGdF6ZuEMt:-KJpnOGYNfyGdF6ZuEMu:-131926322) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L786)):*
> ```XML
>         h2Hash = crypto.createHash('sha256'),
>         h3Hash = crypto.createHash('sha256');
>     h0Hash.update((new Date()).toISOString() + '--' + KeyVerify.seqno);
> ```

Is there really no randomness in this design?  It's hard to believe that this achieves the security goal without a random seed.

---

*[src/generic_core/key-verify.ts, line 839 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpplNE36zAoACwxTrR:-KJpplNE36zAoACwxTrS:1577530593) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L839)):*
> ```XML
>   }
> 
>   private calculateSAS_() :Promise<number> {
> ```

Deserves a comment indicating 16 bits.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2459)*
<!-- Sent from Reviewable.io -->

trevj commented:
Review status: 0 of 18 files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[src/generic_core/key-verify.ts, line 1 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlsYzn4Pqz5uOV6oDX:-KJpwQ4Cp1uABu_KULxI:-644060116) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L1)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Do we know how much stuff this pulls into the uProxy build?  If it's just the ciphers we need that's fine, but if it's also RSA and lots of other stuff ... might be better to use a purpose-built hash package.  Also, we're already importing an md5 package for the identicons, so if this introduces another md5 implementation then we should consider getting rid of the other one.
</details>
We're already using `crypto` in a bunch of places so I think this is fine and, yes, we can replace our use of the `md5` package:
https://github.com/uProxy/uproxy/issues/2464

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2459)*
<!-- Sent from Reviewable.io -->

mollyling commented:
Review status: 0 of 19 files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[src/generic_core/key-verify.ts, line 1 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlsYzn4Pqz5uOV6oDX:-KKFUnumszKvJLYQ0UXj:-310385951) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L1)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> We're already using `crypto` in a bunch of places so I think this is fine and, yes, we can replace our use of the `md5` package:
> https://github.com/uProxy/uproxy/issues/2464
</details>
Acknowledged.

---

*[src/generic_core/key-verify.ts, line 12 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlrOfgWW6OF05Weduo:-KK9iaFfWq4TTAJKcNIJ:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L12)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Can you make this `Object` instead of `any`?
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 16 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlrZ-_VuRd8yQ7I42i:-KK9QK_XxVxLAHWq13U6:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L16)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> resolve
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 27 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlrtYoa7vmKGRcIMtL:-KK9QNjdwm7JWXGmeGqd:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L27)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Space between `)` and `{`
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 31 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlsAeKBQQ0f159WpNL:-KK9QPmY632BgLSvmGG2:-1953493427) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L31)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Could `type` be a `Type` here instead of a string?
</details>
It could, but that'd just make it harder to debug and to (if we ever needed) to extend.  These types are directly JSON'd into wire messages.

---

*[src/generic_core/key-verify.ts, line 52 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlsE4oVLoKd5EjaNPz:-KK9R-S5UMPBGteDKNCG:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L52)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Could you make these types subtypes of a `Message` supertype?
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 102 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJluQ49hGoUSLKfIomY:-KK9RXU_Jg9_zhG2NEc0:1049690553) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L102)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Did you see `buf.toString('hex')`? https://nodejs.org/api/buffer.html#buffer_buffers_and_character_encodings
</details>
Good point!  I used it a few times, but got too fixated on putting spaces between bytes.  So hexify() now gone and the call sites invoke toString('hex').  For larger buffers I'll still use logBuffer, as toString('hex') is rather hard to read for them.

---

*[src/generic_core/key-verify.ts, line 107 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlu8M36RAsG1jxgGJu:-KK9RVwuxmmqZR37SwDf:807670543) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L107)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Why not `buf.readUInt8`?
</details>
Good point!

---

*[src/generic_core/key-verify.ts, line 111 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJltywtqnrhU-4Ky9J3:-KK9SEj2VfUa4WK6qaeq:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L111)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Prefer singlequotes
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 122 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpfoK3nw4ZVov2Gnfh:-KK9i0rCKeKx38qjI7_O:-1686805532) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L122)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> I don't understand everything that's going on here, but it seems like overkill and duplication of code.
</details>
Duplicate in hexify() taken out, and a branch removed after a ->readUInt8() transition.

---

*[src/generic_core/key-verify.ts, line 148 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpg0V7eZKtTEp530JG:-KK9iA2ODRhl7meWiaLu:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L148)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> No space between `( (`
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 158 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpqFryig6BPenrPTFD:-KK9nCl2Wp5c5sYxYSVu:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L158)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> This usage model seems pretty weird.  Maybe it would make more sense to have a pair of factory methods for the two roles?
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 166 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpgN68klkP6PZ7vDCC:-KK9nCMabNFWl020mQPB:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L166)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> `queuedGenerations_` in our code style
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 171 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpgTfJwGIFvW_lc_XK:-KK9vsxhA5w6d3iDyUN3:-279521878) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L171)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> I'd prefer an enum (or a boolean)
</details>
Done. Replaced with a boolean meaning role_ == 0.

---

*[src/generic_core/key-verify.ts, line 192 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJph1wa4JrBuNFrgveJ:-KK9npUmVwWm7R9cmwgA:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L192)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> I think our code style is `PGP_PASSWORD` for constants.
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 198 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJphGwy3cc0aX76aczq:-KK9nq3hmSCV4He53ilB:1634310268) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L198)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> `keyMap_` and `prereqMap_` need descriptions.  I presume you're encoding some sort of state machine?  A comma-separated list of ... something ... encoded as a string, is pretty strange.
</details>
I've added a comment.  I'm happy to discuss better ways to do this.

---

*[src/generic_core/key-verify.ts, line 199 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpiLb84Or9PzV-Glqt:-KK9yHUkbtR17bEe-bs2:1696147916) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L199)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Instead of repeating string literals, I'd prefer having these keys in some sort of variable or collection, just so we aren't retyping each string in each place it's used and risking a typo.
</details>
Ok, replaced with some slight trickery (see new initKeyMap_ static method).

---

*[src/generic_core/key-verify.ts, line 238 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJphWgLzvQ1l5P08_A7:-KK9o6XeKbVSD2qXtCHk:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L238)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Can collapse this to `private peerPubKey_:string`.  Same with `delegate`.
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 297 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpjyERT-jbOFBAtR2A:-KK9w9EUf-PA9ZoRr686:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L297)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> I find this really hard to read.  I think it would be clearer if each message processing algorithm were split out into its own method, so this method just contains control flow.
> 
> I understand why you're trying to keep both sides' flow in a single linear sequence here, but it's fairly counterintuitive to me.  Hopefully it'll be more comprehensible with the detailed logic moved out into helper methods.
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 307 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpjIVhwn76qR6_5fOt:-KK9rbwQjRe5UVCJZGU3:-489111174) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L307)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> We prefer `===`
</details>
Changed into a table lookup

---

*[src/generic_core/key-verify.ts, line 412 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpjbEZwqLOmCUPEzBp:-KKFUhdJAKxAb5uC1uY6:-497587213) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L412)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> typo?
</details>
Fexed

---

*[src/generic_core/key-verify.ts, line 745 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpmrguNCdN6a8tFWF4:-KK9s06kkpSpjl_las4S:-46326288) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L745)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Commented out code?   Also typo.
</details>
Commented out to act as documentation.  Admittedly rather unclear, so fixed with better comment.

---

*[src/generic_core/key-verify.ts, line 786 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpnOGYNfyGdF6ZuEMt:-KK9swDcVUk6g6duMFcU:1984379610) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L786)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Is there really no randomness in this design?  It's hard to believe that this achieves the security goal without a random seed.
</details>
Nope that's a lazy placeholder I forgot to replace with good stuff. Fixed.

---

*[src/generic_core/key-verify.ts, line 839 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpplNE36zAoACwxTrR:-KK9veZFCehR5k55FFEC:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L839)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Deserves a comment indicating 16 bits.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2459)*
<!-- Sent from Reviewable.io -->

mollyling commented:
I've moved all table initialization into the style we discussed in person.

bemasc commented:
Review status: 0 of 19 files reviewed at latest revision, 15 unresolved discussions.

---

*[src/generic_core/key-verify.ts, line 52 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlsE4oVLoKd5EjaNPz:-KKFWfwg4cOUkcUU7_VQ:-2009451043) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L52)):*
<details><summary><i>Previously, mollyling (Molly Ling) wrote…</i></summary>
> Done.
</details>
I think it makes sense to move the `type` attribute up into `Message`.

---

*[src/generic_core/key-verify.ts, line 158 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpqFryig6BPenrPTFD:-KKKAG0TDTte-DcncmON:409377935) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L158)):*
<details><summary><i>Previously, mollyling (Molly Ling) wrote…</i></summary>
> Done.
</details>
Is it?  If so, please update the comment.

---

*[src/generic_core/key-verify.ts, line 785 \[r7\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKK9YP3Pv1GXI0p1lQK:-KKK9YP3Pv1GXI0p1lQL:1603408522) ([raw file](https://github.com/uproxy/uproxy/blob/2584a6f900b157e0325a9e1f7122fe204c79b1c0/src/generic_core/key-verify.ts#L785)):*
> ```XML
>     return this.calculateS0_().then( (s0 :Buffer) => {
>       let h0 = this.ourHashes_.h0;
>       return Promise.resolve<Messages.Tagged>(
> ```

No need for `Promise.resolve` inside any Promise chain.

---

*[src/generic_core/key-verify.ts, line 811 \[r7\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKK8sJ-TAJ0LvUcr4O4:-KKK8sJ0wzCNIk6ZRrsK:-1706379850) ([raw file](https://github.com/uproxy/uproxy/blob/2584a6f900b157e0325a9e1f7122fe204c79b1c0/src/generic_core/key-verify.ts#L811)):*
> ```XML
>         h3Hash = crypto.createHash('sha256');
>     h0Hash.update(crypto.randomBytes(256 / 8));
>     let h0 = h0Hash.digest();
> ```

In the ZRTP spec H0 is a 256-bit nonce, not a hash (https://tools.ietf.org/html/rfc6189#section-9), so this is different from the spec.  It's not "wrong", but is there a reason to deviate?

---

*[src/generic_core/key-verify.ts, line 858 \[r7\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKK9SnpDXa2KiPZszA-:-KKK9SnpDXa2KiPZszA0:-748957394) ([raw file](https://github.com/uproxy/uproxy/blob/2584a6f900b157e0325a9e1f7122fe204c79b1c0/src/generic_core/key-verify.ts#L858)):*
> ```XML
>         logBuffer('s0', s0);
>         this.s0_ = s0;
>         return Promise.resolve<Buffer>(s0);
> ```

No need for `Promise.resolve`.

---

*[src/generic_core/key-verify.ts, line 885 \[r7\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKK9JWdxdaIfQVzz1Sr:-KKK9JWdxdaIfQVzz1Ss:1126649453) ([raw file](https://github.com/uproxy/uproxy/blob/2584a6f900b157e0325a9e1f7122fe204c79b1c0/src/generic_core/key-verify.ts#L885)):*
> ```XML
>         logBuffer('sasvalue', sasvalue);
>         let sasHumanInt :number = sasvalue.slice(0,2).readUInt16BE(0);
>         return Promise.resolve<number>(sasHumanInt);
> ```

You can just return the number.  No need for `Promise.resolve`.

---

*[src/generic_core/key-verify.ts, line 923 \[r7\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKK7lghOj_z4m_9bqcD:-KKK7lghOj_z4m_9bqcE:-240167218) ([raw file](https://github.com/uproxy/uproxy/blob/2584a6f900b157e0325a9e1f7122fe204c79b1c0/src/generic_core/key-verify.ts#L923)):*
> ```XML
>     let fourth_hash_buf = unbase64Concat(dhpart2.mac);
>     let total_hash_buf = Buffer.concat(
>       [].concat( first_hash_buf, cv_buf, second_hash_buf, dh1k_buf,
> ```

Could this just be `[first_hash_buf, ... fourth_hash_buf]`?

---

*[src/generic_core/key-verify.ts, line 968 \[r7\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKK6MA9mgivt9_7YbP7:-KKK6MA9mgivt9_7YbP8:584418731) ([raw file](https://github.com/uproxy/uproxy/blob/2584a6f900b157e0325a9e1f7122fe204c79b1c0/src/generic_core/key-verify.ts#L968)):*
> ```XML
> 
>   // key and value are both base64-encoded.
>   private fullHmac_(key:string, value:string) :Buffer {
> ```

Could you use `crypto.createHmac` instead?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2459)*
<!-- Sent from Reviewable.io -->

mollyling commented:
Review status: 0 of 19 files reviewed at latest revision, 15 unresolved discussions.

---

*[src/generic_core/key-verify.ts, line 52 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJlsE4oVLoKd5EjaNPz:-KKKBfqorcikXNfandqG:738682207) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L52)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> I think it makes sense to move the `type` attribute up into `Message`.
</details>
I want to keep the `Message` type the same as the wire protocol message, and keep any other metadata we care about in a separate type (`Tagged`).

---

*[src/generic_core/key-verify.ts, line 158 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJpqFryig6BPenrPTFD:-KKKHUovrmLJKWrl6gpC:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L158)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Is it?  If so, please update the comment.
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 811 \[r7\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKK8sJ-TAJ0LvUcr4O4:-KKKGmfYy8hlhGCL8SGf:1758359711) ([raw file](https://github.com/uproxy/uproxy/blob/2584a6f900b157e0325a9e1f7122fe204c79b1c0/src/generic_core/key-verify.ts#L811)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> In the ZRTP spec H0 is a 256-bit nonce, not a hash (https://tools.ietf.org/html/rfc6189#section-9), so this is different from the spec.  It's not "wrong", but is there a reason to deviate?
</details>
Fixed.

---

*[src/generic_core/key-verify.ts, line 923 \[r7\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKK7lghOj_z4m_9bqcD:-KKKH7TblFTeEg-kOUHT:2067975471) ([raw file](https://github.com/uproxy/uproxy/blob/2584a6f900b157e0325a9e1f7122fe204c79b1c0/src/generic_core/key-verify.ts#L923)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Could this just be `[first_hash_buf, ... fourth_hash_buf]`?
</details>
I tried that first, found that the browserified Buffer.concat() didn't work with that form.  I forget why.

---

*[src/generic_core/key-verify.ts, line 968 \[r7\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKK6MA9mgivt9_7YbP7:-KKKH3I0yYcIbzoUq38p:-616780) ([raw file](https://github.com/uproxy/uproxy/blob/2584a6f900b157e0325a9e1f7122fe204c79b1c0/src/generic_core/key-verify.ts#L968)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Could you use `crypto.createHmac` instead?
</details>
See comment below w.r.t. FIPS-198.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2459)*
<!-- Sent from Reviewable.io -->

mollyling commented:
Thanks.  Anything else?

bemasc commented:
Review status: 0 of 19 files reviewed at latest revision, 21 unresolved discussions.

---

*[src/generic_core/key-verify.ts, line 198 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJphGwy3cc0aX76aczq:-KKPOxFwO1GsjlJsFuMD:-565697684) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L198)):*
<details><summary><i>Previously, mollyling (Molly Ling) wrote…</i></summary>
> I've added a comment.  I'm happy to discuss better ways to do this.
</details>
This comment should probably mention JSON now.

---

*[src/generic_core/key-verify.ts, line 811 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPP_oenPgLVWIL81pt:-KKPP_oenPgLVWIL81pu:-1777003107) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/key-verify.ts#L811)):*
> ```XML
>   private generateHashes_() :Hashes {
>     var result : Hashes = new Hashes();
>     let h0Hash = crypto.createHash('sha256'),
> ```

`h0Hash` is unused now.

---

*[src/generic_core/key-verify.ts, line 847 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPPdmyeFi1Mxhz2FcO:-KKPPdmzRV3Z9hr_wp-g:1460037271) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/key-verify.ts#L847)):*
> ```XML
>         beZero.fill(0);
>         beOne.writeInt32BE(1,0);
>         
> ```

Nit: trailing spaces.

---

*[src/generic_core/remote-instance.ts, line 299 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPQsMz-feaZqxjv3MJ:-KKPQsMz-feaZqxjv3MK:-2029522695) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/remote-instance.ts#L299)):*
> ```XML
>       if (this.verifyState_ === social.VerifyState.VERIFY_COMPLETE) {
>         log.debug('verifyUser(%1): ALREADY VERIFIED.', firstMsg);
> //        return;
> ```

Commented out code?

---

*[src/generic_core/remote-instance.ts, line 303 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPRwfkH1Y5_yjaqq9s:-KKPRwfkH1Y5_yjaqq9t:-953393552) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/remote-instance.ts#L303)):*
> ```XML
>       let inst = this;
>       let clientId = this.user.instanceToClient(this.instanceId);
>       let delegate = <key_verify.Delegate>{
> ```

Can you do `let delegate :key_verify.Delegate = {` instead?  That would make it clearer that the interface type is being enforced, not casted.

---

*[src/generic_core/remote-instance.ts, line 317 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPYpu1U_VIj2qipZwj:-KKPYpu1U_VIj2qipZwk:-1095573906) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/remote-instance.ts#L317)):*
> ```XML
>             // Send UPDATE message with SAS.
>             this.resolvedVerifySAS_ = resolve;
>             // The UI's now showing the SAS with a YES/NO prompt.  The
> ```

Nit: There's no need for this to be inside the Promise constructor, so it might as well be afterward. (Minimum indent principle.)

---

*[src/generic_core/uproxy_core.ts, line 804 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPV735Z887PUafUG0m:-KKPV736n8cqI9hCipZJ:284607660) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/uproxy_core.ts#L804)):*
> ```XML
>   }
> 
>   // Figure out what the returned promise is for.  Probably nothing.
> ```

Please replace this comment with the answer to this question.

---

*[src/generic_core/uproxy_core.ts, line 806 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPTT45JsZkFpz-DFKw:-KKPTT468_RRdAVDwf7z:-778384883) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/uproxy_core.ts#L806)):*
> ```XML
>   // Figure out what the returned promise is for.  Probably nothing.
>   public verifyUser = (inst:social.InstancePath) :Promise<void> => {
>     console.log('app.core: verifyUser:', inst);
> ```

Looks like `console.log` is not the recommended way to log in this file.

---

*[src/generic_core/uproxy_core.ts, line 807 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPX0Bt-2fHhfDs4_6q:-KKPX0Bt-2fHhfDs4_6r:1246246512) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/uproxy_core.ts#L807)):*
> ```XML
>   public verifyUser = (inst:social.InstancePath) :Promise<void> => {
>     console.log('app.core: verifyUser:', inst);
>     var network = <social_network.AbstractNetwork>this.getNetworkByName_(
> ```

Why is this typecast required?  At a glance it looks like it ought to work without this.

---

*[src/generic_ui/polymer/instance.html, line 69 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPYBBLbkUBIbzaV-sy:-KKPYBBM2Hn5lHq5DtL4:1401277667) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_ui/polymer/instance.html#L69)):*
> ```HTML
>         <span class='verified'
>               hidden?='{{ instance.verifyState != VerifyState.VERIFY_COMPLETE }}'>
>           Verified </span>
> ```

Needs i18n?  Also, whitespace seems inconsistent with the rest of this file.

---

*[src/generic_ui/polymer/instance.ts, line 61 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPXSIhYWBbsBnrEEmv:-KKPXSIiesA-W49pPLEq:867755173) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_ui/polymer/instance.ts#L61)):*
> ```XML
>   },
>   verify: function() {
>     if (this.instance.verifyState != social.VerifyState.VERIFY_BEGIN) {
> ```

Nit: `!==`

---

*[src/generic_ui/scripts/ui.ts, line 1431 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPVZIKD1vC2KHWaSCC:-KKPVZIKD1vC2KHWaSCD:-802028230) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_ui/scripts/ui.ts#L1431)):*
> ```XML
>   public finishVerifying = (inst :social.InstanceData,
>                             sameSAS: boolean) :Promise<void> => {
>     var args = <uproxy_core_api.FinishVerifyArgs> {
> ```

Can this be `let args :uproxy_core_api.FinishVerifyArgs = {`?

---

*[src/interfaces/uproxy_core_api.ts, line 52 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPVH6Ogu9sPVzBVdYY:-KKPVH6Ogu9sPVzBVdYZ:-348027158) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/interfaces/uproxy_core_api.ts#L52)):*
> ```XML
> }
> 
> export const kFeatureVerify = 'verify';
> ```

Nit: `FEATURE_VERIFY`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2459)*
<!-- Sent from Reviewable.io -->

mollyling commented:
Review status: 0 of 19 files reviewed at latest revision, 21 unresolved discussions, some commit checks pending.

---

*[src/generic_core/key-verify.ts, line 198 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KJphGwy3cc0aX76aczq:-KKPecPLiC6V57wG0Vpy:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/f9dde62380101f5113970209b3c424f1b9d2ca00/src/generic_core/key-verify.ts#L198)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> This comment should probably mention JSON now.
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 785 \[r7\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKK9YP3Pv1GXI0p1lQK:-KKPf4-dAVWQHBDnzo8W:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/2584a6f900b157e0325a9e1f7122fe204c79b1c0/src/generic_core/key-verify.ts#L785)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> No need for `Promise.resolve` inside any Promise chain.
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 858 \[r7\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKK9SnpDXa2KiPZszA-:-KKPewyBs0PggBZka1uG:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/2584a6f900b157e0325a9e1f7122fe204c79b1c0/src/generic_core/key-verify.ts#L858)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> No need for `Promise.resolve`.
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 885 \[r7\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKK9JWdxdaIfQVzz1Sr:-KKPexHHXmodvE1vq5lk:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/2584a6f900b157e0325a9e1f7122fe204c79b1c0/src/generic_core/key-verify.ts#L885)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> You can just return the number.  No need for `Promise.resolve`.
</details>
Done.

---

*[src/generic_core/key-verify.ts, line 811 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPP_oenPgLVWIL81pt:-KKPeose8LxJJkDeOl_c:610327451) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/key-verify.ts#L811)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> `h0Hash` is unused now.
</details>
Removed

---

*[src/generic_core/remote-instance.ts, line 299 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPQsMz-feaZqxjv3MJ:-KKPf6fCsPVOy7tEHgPy:-568479996) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/remote-instance.ts#L299)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Commented out code?
</details>
While testing the UI, it's out so I can re-run the verification step. But the UI can just hide the button instead of early-exiting here.

---

*[src/generic_core/remote-instance.ts, line 303 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPRwfkH1Y5_yjaqq9s:-KKPfTjX2LVwmxESAEE-:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/remote-instance.ts#L303)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Can you do `let delegate :key_verify.Delegate = {` instead?  That would make it clearer that the interface type is being enforced, not casted.
</details>
Done.

---

*[src/generic_core/remote-instance.ts, line 317 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPYpu1U_VIj2qipZwj:-KKPfhanzWNFkiz9c9Tm:-1863572015) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/remote-instance.ts#L317)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: There's no need for this to be inside the Promise constructor, so it might as well be afterward. (Minimum indent principle.)
</details>
This avoids updating state to the UI with the SAS before we're able to handle it here (with that resolution function).  Otherwise there's a race.

---

*[src/generic_core/uproxy_core.ts, line 804 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPV735Z887PUafUG0m:-KKPiA3xJ6v_efxs9GP_:224603748) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/uproxy_core.ts#L804)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Please replace this comment with the answer to this question.
</details>
I still don't know it -- I was copying the signature from other methods like `removeContact()` and others in `CoreApi`.  Removed.

---

*[src/generic_core/uproxy_core.ts, line 806 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPTT45JsZkFpz-DFKw:-KKPgIxu4TjASBb_fEhn:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/uproxy_core.ts#L806)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Looks like `console.log` is not the recommended way to log in this file.
</details>
Done.

---

*[src/generic_core/uproxy_core.ts, line 807 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPX0Bt-2fHhfDs4_6q:-KKPgJblsS1SzddDca4V:7195195) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_core/uproxy_core.ts#L807)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Why is this typecast required?  At a glance it looks like it ought to work without this.
</details>
The declared return type of that function is a superclass of `AbstractNetwork` at the API level, but an instance of `AbstractNetwork` on  implementation.  The latter is necessary.  This was found from a discussion with dborkan@

---

*[src/generic_ui/polymer/instance.html, line 69 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPYBBLbkUBIbzaV-sy:-KKPgZC6Ns2nIPzjNoKh:-1009909484) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_ui/polymer/instance.html#L69)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Needs i18n?  Also, whitespace seems inconsistent with the rest of this file.
</details>
Needs a real UI implementation.  But I'll wrap it in i18n.  I'll ask in person about whitespacing.

---

*[src/generic_ui/polymer/instance.ts, line 61 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPXSIhYWBbsBnrEEmv:-KKPgl_Lg4uGfJxhxoAy:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_ui/polymer/instance.ts#L61)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: `!==`
</details>
Done.

---

*[src/generic_ui/scripts/ui.ts, line 1431 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPVZIKD1vC2KHWaSCC:-KKPhqGO2JwMTQTh0eKK:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/generic_ui/scripts/ui.ts#L1431)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Can this be `let args :uproxy_core_api.FinishVerifyArgs = {`?
</details>
Done.

---

*[src/interfaces/uproxy_core_api.ts, line 52 \[r12\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKPVH6Ogu9sPVzBVdYY:-KKPi-hR_XjZO15hiuDR:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/133a98165571bd4f2e565f6dd3a48140bb7a15a6/src/interfaces/uproxy_core_api.ts#L52)):*
<details><summary><i>Previously, bemasc (Benjamin M. Schwartz) wrote…</i></summary>
> Nit: `FEATURE_VERIFY`. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2459)*
<!-- Sent from Reviewable.io -->

mollyling commented:
Thanks, PTAL.

dborkan commented:
I'm curious what happens if something goes wrong during verification, for example either:
* I start verification (VERIFY_BEGIN state), but my peer and I both quit uProxy before confirming or rejecting the SAS?  (I suspect this will work because the SAS is written to storage but just want to check)
* I start verification, but for some reason one of the messages isn't delivered (maybe something unexpectedly goes wrong with my connection or with the social network).  I suspect maybe this will leave uProxy in a bad state (because my local storage will remember that I'm in a VERIFY_BEGIN state with the remote instance, but no SAS will be saved)?  Maybe if we have state as VERIFY_BEGIN but no SAS available we should reset to VERIFY_NONE?

---

Review status: 0 of 19 files reviewed at latest revision, 25 unresolved discussions.

---

*[src/generic_core/key-verify.ts, line 82 \[r13\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKU6NhxwwcpBLkPvB1f:-KKU6Nhy5mPw5BpHJ9m0:-1996639123) ([raw file](https://github.com/uproxy/uproxy/blob/cb84426e5f4e68b43fdd35968bf2d46c4989da28/src/generic_core/key-verify.ts#L82)):*
> ```XML
> // Hash facilities
> //
> export class HashPair {
> ```

Do HashPair and Hashes need to be exported?  Seems like they aren't used outside of this file

---

*[src/generic_core/remote-instance.ts, line 15 \[r13\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKQ3Mvg_ByysFmDeHc3:-KKQ3Mvg_ByysFmDeHc4:262423681) ([raw file](https://github.com/uproxy/uproxy/blob/cb84426e5f4e68b43fdd35968bf2d46c4989da28/src/generic_core/remote-instance.ts#L15)):*
> ```XML
> import consent = require('./consent');
> import crypto = require('./crypto');
> import datachannel = require('../lib/webrtc/datachannel');
> ```

NIT: looks like datachannel is unused in this file

---

*[src/generic_core/remote-instance.ts, line 304 \[r13\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKQ5lfAYOBM-EwWHHOj:-KKQ5lfAYOBM-EwWHHOk:655690437) ([raw file](https://github.com/uproxy/uproxy/blob/cb84426e5f4e68b43fdd35968bf2d46c4989da28/src/generic_core/remote-instance.ts#L304)):*
> ```XML
>       let delegate :key_verify.Delegate = {
>         sendMessage : (msg:any) :Promise<void> => {
>           let instanceMessage :social.PeerMessage = {
> ```

Nit: can you change this name to something like "verifyMessage"?  instanceMessage usually means the INSTANCE_HANDSHAKE type (with permission info), which is different than this

---

*[src/generic_ui/polymer/instance.ts, line 57 \[r13\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKU1bM-KibQDvgGL35M:-KKU1bM0kkktL2B7ipHB:-2005731094) ([raw file](https://github.com/uproxy/uproxy/blob/cb84426e5f4e68b43fdd35968bf2d46c4989da28/src/generic_ui/polymer/instance.ts#L57)):*
> ```XML
>     this.fire('instance-changed');
>   },
>   sasUpdated: function() {
> ```

It seems like you could just use ```instance.verifySAS``` in the .html template, then you might not need the copy of this.sas or the observe function?  Or do you need a copy so that you can set it to null in confirmSas + rejectSas without affecting the instance object?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-:-KKU3IvHL1o77BeG3dgx:-1459328302)*
<!-- Sent from Reviewable.io -->

mollyling commented:
1) The verification state saved is different than the enum.  We just save whether verification has ever succeeded, and init the enum to ether VERIFY_COMPLETE or VERIFY_NONE.

2) That's a danger here. I've added a timeout.  The scaffold UI can't cancel itself, but that'll get sorted in the real thing.  Behind the scenes, the app state properly cancels on both sides after 2 minutes.

---

Review status: 0 of 19 files reviewed at latest revision, 25 unresolved discussions.

---

*[src/generic_core/key-verify.ts, line 82 \[r13\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKU6NhxwwcpBLkPvB1f:-KKiVfhvXsGuP4TbMmUe:-1377089141) ([raw file](https://github.com/uproxy/uproxy/blob/cb84426e5f4e68b43fdd35968bf2d46c4989da28/src/generic_core/key-verify.ts#L82)):*
<details><summary><i>Previously, dborkan wrote…</i></summary>
> Do HashPair and Hashes need to be exported?  Seems like they aren't used outside of this file
</details>
They're part of the constructor for KeyVerify, set up so that we can test it against known inputs.

---

*[src/generic_core/remote-instance.ts, line 15 \[r13\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKQ3Mvg_ByysFmDeHc3:-KKiVnZKBRSlAf0JLptt:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/cb84426e5f4e68b43fdd35968bf2d46c4989da28/src/generic_core/remote-instance.ts#L15)):*
<details><summary><i>Previously, dborkan wrote…</i></summary>
> NIT: looks like datachannel is unused in this file
</details>
Done.

---

*[src/generic_core/remote-instance.ts, line 304 \[r13\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKQ5lfAYOBM-EwWHHOj:-KKiVtOc4b_wE1QrsvdU:-485110604) ([raw file](https://github.com/uproxy/uproxy/blob/cb84426e5f4e68b43fdd35968bf2d46c4989da28/src/generic_core/remote-instance.ts#L304)):*
<details><summary><i>Previously, dborkan wrote…</i></summary>
> Nit: can you change this name to something like "verifyMessage"?  instanceMessage usually means the INSTANCE_HANDSHAKE type (with permission info), which is different than this
</details>
Sure.

---

*[src/generic_ui/polymer/instance.ts, line 57 \[r13\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKU1bM-KibQDvgGL35M:-KKUBGExO_k-8hdGlvyu:-858657831) ([raw file](https://github.com/uproxy/uproxy/blob/cb84426e5f4e68b43fdd35968bf2d46c4989da28/src/generic_ui/polymer/instance.ts#L57)):*
<details><summary><i>Previously, dborkan wrote…</i></summary>
> It seems like you could just use ```instance.verifySAS``` in the .html template, then you might not need the copy of this.sas or the observe function?  Or do you need a copy so that you can set it to null in confirmSas + rejectSas without affecting the instance object?
</details>
The latter.  I can put in a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-:-KKiVAKbi4rONoQnebMx:1726696929)*
<!-- Sent from Reviewable.io -->

mollyling commented:
Hi, please let me if there's anything else.  Thanks.

bemasc commented:
:+1: From me.

---

Reviewed 1 of 1 files at r15.
Review status: 0 of 19 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[src/generic_core/key-verify.ts, line 191 \[r14\]](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-KKjGKcVCvnkMKwr5YFF:-KKjGKcW4v1Isl-KVm9K:1850646821) ([raw file](https://github.com/uproxy/uproxy/blob/ee0480b5ef8a2245b70f64a4bb16dd4a8a4cd69d/src/generic_core/key-verify.ts#L191)):*
> ```XML
>   private pgp_ :freedom.PgpProvider.PgpProvider;
> 
>   // Data that requires expensive calculations to make.  These are a innately
> ```

Nit: "a innately"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-:-KKjbTuQunMkxj_2q2-4:166706081)*
<!-- Sent from Reviewable.io -->

dborkan commented:
:+1: 

---

Review status: 0 of 19 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2459#-:-KKjbgbK96jQeTWdmUKc:-510466975)*
<!-- Sent from Reviewable.io -->
